src/jit/hwintrinsiccodegenxarch.cpp

```diff
@@ -94,7 +94,8 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
 
             case 2:
                 genConsumeOperands(node);
-                if (ival != -1)
+                if ((ival != -1) && (((baseType == TYP_DOUBLE) && (flags & HW_Flag_Imm_Double)) ||
+                                     ((baseType == TYP_FLOAT) && (flags & HW_Flag_Imm_Float))))
```

Comment:
Looked the code a bit more. I see the problem is 
1. just some of float/double compare-instructions need the fixed imm values and integral comparisons never require it.
2. `Sse2.CompareLessThan` does not have the direct instructions on integral types

So, IMO, for problem (1) here can be 
```cpp
if ((ival != -1) && varTypeIsFloating(baseType))
{ ...
```
For (2), we just need to specially treat `Sse2.CompareLessThan`.

Is there any other intrinsic would break the above rules?

Reply:
> Is there any other intrinsic would break the above rules?

Yes, I think that @tannergooding has indicated in comments above that there are several floating point instructions which would require reversing operand order - see https://github.com/dotnet/coreclr/pull/15585#discussion_r163067471

Reply:
> Yes, I think that @tannergooding has indicated in comments above that there are several floating point instructions

I think @tannergooding was talking about another story...

Reply:
> was talking about another story...

Yes. We have some operands which are commutative but only if you also inverse the operation being done. I was indicating we want a flag to track that.

But that for the general case of knowing whether or not the fixed ival can be used, we should just be able to do `if ((ival != -1) && varTypeIsFloating(baseType))`

Reply:
> But that for the general case of knowing whether or not the fixed ival can be used, we should just be able to do if ((ival != -1) && varTypeIsFloating(baseType))

This logic will accept 'cmppd` with `float`, and `cmpps` with `double` if `ival != -1`

Reply:
> Sse2.CompareLessThan case

We should expect it to be repeated for every intrinsic with this functionality - 4 - 6 of them total for all current ISAs.

Reply:
Those cases should have already been handled by the importer and caused an invalid result. It is also trivial enough to check with an assert.

Reply:
Will fix it.

Reply:
@fiigii @tannergooding Final opinion on `HW_Flags_ReverseIntegralOperands` flag - should I scrap this flag and go for special case handling or keep for 4 - 6 intrinsics total.

Reply:
> This logic will accept cmppd with float, and cmpps with double if ival != -1

I do not think so. Once you set the correct instructions in the table, it would never happen. 

> We should expect it to be repeated for every intrinsic with this functionality - 4 - 6 of them total for all current ISAs.

Is there other intrinsic that has to switch operands? I did not find in the manual.

Reply:
Yes. Intrinsics are missing in AVX2 and SSE4_1, but underlying instructions exists - IMO we should support basic comparison operators which are: `<`, `>`, `<=`, `>=`, `==`, `!=` for all types - but in AVX, AVX2 there are no intrinsics covering similar range of types and operations like in SSE, SSE2.

SSE4_1
```
PCMPGTQ xmm1,xmm2/m128

```

AVX2
```
VPCMPGTB ymm1, ymm2, ymm3/m256
VPCMPGTW ymm1, ymm2, ymm3/m256
VPCMPGTD ymm1, ymm2, ymm3/m256
VPCMPGTQ ymm1, ymm2, ymm3/m256

```
AVX512
```
VPCMPGTQ k1 {k2}, zmm2, zmm3/m512/m64bcst
```

In AVX/AVX2 we have only `CompareEqual` and `CompareGreaterThan` intrinsics defined while IMO we should have if possible `<`, `>`, `<=`, `>=`, `==`, `!=`. But this another problem for separate discussion. SSE4_1 is missing `CompareGreaterThan`  for `PCMPGTQ` and reversed operands equivalent `CompareLessThan`.

Reply:
> IMO we should support basic comparison operators 

Ah, I see. In the current HW intrisnic API design, we are trying to 
- match C++ counterpart that hlpes users to port their applications from C++ to .NET Core
- and provoid as less as possible hepler functions like `Sse2.CompareLessThan`. So, I do not think it is necessary to provide these helper functions in the low-level APIs.

Additinally, even though we want these helpers in the future, adding a flag `HW_Flags_ReverseIntegralOperands` is not a good idea. We should transform the helpers to the actual intrinsic IR in the importer. For example, `Sse2.CompareLessThan(op1, op2)` should be converted `Sse2.CompareGreaterThan(op2, op1)`, which would simplify the subsequent optimization work (i.e., VN and CSE).

Reply:
Will remove flag and handle it as a helper.

Anyway SSE4_1 is missing CompareGreaterThan for PCMPGTQ even without operand reversal.

Reply:
> SSE4_1 is missing CompareGreaterThan for PCMPGTQ 

This is a SSE4.2 intrinsic, and we have it.